### PR TITLE
Fix #597

### LINKED ISSUE #597: Patch for broken ICC profile in PNG files
We have been notified of a problem with digiKam writing metadata to PNG files, after which the ICC profile is broken.

https://bugs.kde.org/show_bug.cgi?id=401970

Exiv2 uses a fixed profile name "ICC PROFILE" with 11 characters + zero separator. The profile name can vary between 1-79 characters. Also a warning of the libpng with wrong ICC profile CRC checksum fixes this patch.

Maik
[pngIccProfileName.txt](https://github.com/Exiv2/exiv2/files/2683451/pngIccProfileName.txt)

### PR DESCRIPTION
**Pull request by piponazo**
Apply the patch provided in #597.

### CODE REVIEW

**Review by clanmills — DISMISSED** [review]
I think this is OK.  The profile name is 1-79 bytes:  https://www.w3.org/TR/PNG/#11iCCP

**Comment by clanmills** [comment]
The test harness is failing because we have changed how we calculate the checksum.  So, we get a warning on every ICC profile checksum.  The easy way to fix this is to modify the reference file test/data/icc-test.out to have the new values.  However that only silences the test harness.  We need to dig deeper to understand what's going on here.

If we have been calculating the checksum incorrectly, it's strange that this has never been reported.  I don't think we have a checksum validator in our code, so an application (such as digiKam) is probably unaware if there's an issue with the checksum.  I believe the ICC Color Consortium publish sample files and I'll investigate Exiftool's handing of this.  

I think the existing code ignored the profile name (the 1-79 byte string), so we never read or wrote that.  I think we'll need to modify *`class Image`* to contain the profile name and when we write the ICC profile, we should rewrite the profile name.

I see there are remarks in the Specification about the permissible bytes in the profile name.   Should we test them?  Probably not.  If the file has violated that part of the specification, we can ignore that.

The command **`$ exiv2 -iC`** (insert color profile) does not write a profile name into the Image.  If writeMetadata() is being called to rewrite an ICC profile which we discovered with `readMetadata()`, we should preserve that name.  If we insert an ICC profile into an image we should invent a name such as `exiv2-date-time`

We should investigate if the profile_name should be stored in other image formats such as JPEG and TIFF. 

The code reading the profile name can read past the end of the buffer.  It assumes there will be a null byte within 79 characters.  However the buffer might be less than 79 bytes.  ASAN will detect that with a suitably fuzzed file.  We'll need to strengthen the code.

There's quite a lot more work ahead here.

**Comment by piponazo** [comment]
Great feedback about the issue @clanmills . We'll investigate it after Xmas. I just decided to apply the patch provided in #597 and check if everything was fine with it. Clearly the change requires more work.

**Comment by clanmills** [comment]
I found this: http://schaik.com/png/pngcsum.html  For sure, we can use that to ensure we're calculating the checksum correctly.  Let's should resist recalculating every checksum in a file as I believe that requires us to read every byte.

And there are lots of PNG files (including corrupted files). http://www.schaik.com/pngsuite/pngsuite.html

**Comment by piponazo** [comment]
> The easy way to fix this is to modify the reference file test/data/icc-test.out to have the new values. However that only silences the test harness. We need to dig deeper to understand what's going on here.

From my point of view that should be enough. We can also try to double check if checksum is correct with the links that you reported in your last comment.

> If we have been calculating the checksum incorrectly, it's strange that this has never been reported. I don't think we have a checksum validator in our code, so an application (such as digiKam) is probably unaware if there's an issue with the checksum.

Maybe @cgilles could provide here some feedback. Is digiKam checking those checksum? I am still relatively new in the project, but for me it only makes sense to check those values at Exiv2 level and not in the consumers of Exiv2. 

I will dig more into this issue once I come back from my holidays.

**Comment by cgilles** [comment]
As i know : 

In digiKam, we don't use Exiv2 to play with ICC profile metadata. We use lcms version 2 API to read and compute Color Management everywhere in application.

digiKam provide a CM control panel where user can setup the icc color profiles to use in workflow (with RAW files and camera profiles) . We also provide some standard icc color profiles available in public domain, as binary files (D65, ProPhoto, AdobeRGB) : 

https://cgit.kde.org/digikam.git/tree/core/data/profiles

About the registration in image files after editing, we delegate the ICC color profile record to file to the image codec, as libjpeg, libpng, libtiff, libjasper, that we depend. There is no checksum computed here. the low level library must do the stuff as well in background.

Look for exemple the ICC profile management in DK image loader : 

PNG read : https://cgit.kde.org/digikam.git/tree/core/libs/dimg/loaders/pngloader.cpp#n518
PNG write: https://cgit.kde.org/digikam.git/tree/core/libs/dimg/loaders/pngloader.cpp#n792

TIFF read : https://cgit.kde.org/digikam.git/tree/core/libs/dimg/loaders/tiffloader.cpp#n263
TIFF write : https://cgit.kde.org/digikam.git/tree/core/libs/dimg/loaders/tiffloader.cpp#n881

JPEG read : https://cgit.kde.org/digikam.git/tree/core/libs/dimg/loaders/jpegloader.cpp#n611
JPEG write : https://cgit.kde.org/digikam.git/tree/core/libs/dimg/loaders/jpegloader.cpp#n848

Gilles

**Comment by clanmills** [comment]
Thanks for the clarification, @cgilles  That's very helpful. Enjoy your vacation, @piponazo 

I think we will:
1) adopt the patch (more or less)
2) store the profile name in class Image and write it in writeMetadata()
3) use the utility  http://schaik.com/png/pngcsum.html to verify our checksum is correct
4) modify the test harness reference file
I don't believe TIFF and JPEG use the "profile name"

So this will get resolved in January.  Happy New Year, everybody.

**Comment by mqualmann** [comment]
A good test is to draw a PNG image with Gimp and to integrate a color profile (sRGB). This image is easily read by digiKam and also the color profile. In the console, libpng reports no error. Now we write a tag in the image. The color profile is preserved because Gimp uses the same profile name (length) as Exiv2. But libpng now reports a CRC error when loading the image.

Maik

**Comment by clanmills** [comment]
@mqualmann That __more or less__ matches my observation.  We'll get this fixed in the next week or two.  Let's relax.  Time for a break.

**Comment by piponazo** [comment]
Hi @clanmills . I spent the evening in this topic by reading the relevant pieces of the PNG specification and reviewing the `PngImage::readMetadata()` function. 

As you mentioned we were ignoring the profile name and the patch provided in #597 was fixing the issue. 

> I think the existing code ignored the profile name (the 1-79 byte string), so we never read or wrote that. I think we'll need to modify class Image to contain the profile name and when we write the ICC profile, we should rewrite the profile name.

Right, we were not reading it properly and now we are storing it with the name "ICC PROFILE". In this line, "ICC PROFILE" is the name, the first "\0" the null separator, and the second "\0" is the default compression.

```cpp
const byte* header   = (const byte*) "ICC PROFILE\0\0" ; // \0 = default compression
```
I could take a step more and keep the profile name in the **PngImage** class to use it later. I am not so sure about keeping it in the **Image** class ... I do not know if it could have side effects.

> I see there are remarks in the Specification about the permissible bytes in the profile name. Should we test them?

I do not think it is needed. If somebody provide creepy images, we will just keep the profile name.

> The code reading the profile name can read past the end of the buffer. It assumes there will be a null byte within 79 characters. However the buffer might be less than 79 bytes.

Where you referring to the old code? The new code does not make that assumption. It tries to find a null separator, but if it is not there it just reads the 79 bytes (or less, depending on the chunkLength). I think the patch provided is well implemented. 

You made some other comments that are still valid, but maybe out of the scope fo #597. We could create another issue in Github and investigate those things later.

File: src/jp2image.cpp [hunk 1]
```diff
@@ -23,7 +23,6 @@
 */
 
 // *****************************************************************************
-// #define DEBUG
```
Comment (piponazo) [inline src/jp2image.cpp]: Note that I get rid of these handmade definitions. We can pass the DEBUG flag through CMake easily

File: src/pngimage.cpp [hunk 2]
```diff
@@ -215,9 +215,7 @@ namespace Exiv2 {
         if (io_->open() != 0) {
             throw Error(kerDataSourceOpenFailed, io_->path(), strError());
         }
-        // Ensure that this is the correct image type
         if (!isPngType(*io_, true)) {
-            if (io_->error() || io_->eof()) throw Error(kerFailedToReadImageData);
```
Comment (piponazo) [inline src/pngimage.cpp]: This check is moved inside `isPngType`

**Review by clanmills — APPROVED** [review]
This looks very good, Luis.  Your new code looks much nicer and simpler.

I ran pngcsum (reference above) on 'master' and your branch.  You've fixed the checksum.  Good Job.

```
764 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/build $ exiv2 -pS ../test/tmp/ReaganLargePng.png  | grep iCCP
    9154 | iCCP  |     293 | ICC PROFILE..x.c``2ptqre.``..+ | 0x5116d227
765 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/build $ pngcsum ../test/tmp/ReaganLargePng.png  /tmp/foo.png | grep iCCP
iCCP (293 ) - csum = 5116d227 -> 5c9e2b23  <--- Exiv2/master has this wrong
766 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2/build $ cd ../../../piponazo/exiv2/build
767 rmills@rmillsmbp:~/gnu/github/piponazo/exiv2/build $ pngcsum ../test/tmp/ReaganLargePng.png  /tmp/foo.png | grep iCCP
iCCP (293 ) - csum = 7d41600b <--- You have fixed this.
768 rmills@rmillsmbp:~/gnu/github/piponazo/exiv2/build $ exiv2 -pS ../test/tmp/ReaganLargePng.png  | grep iCCP
    9154 | iCCP  |     293 | ICC profile..x.c``2ptqre.``..+ | 0x7d41600b <--- You have fixed this.
769 rmills@rmillsmbp:~/gnu/github/piponazo/exiv2/build $ 
```

**Not thread safe** (I am to blame)
static bool bFirst = true;

The static int value[256] vector is fine.  However, we should generate the code to initialise it at compile time.  Then we can remove static bFirst and hex digits. 

**Cosmetic Suggestion**
There's a lot of 
```
#if DEBUG
    std::cerr << bla bla
#endif
```
3 lines of code.  Can we condense that to a single line?  A macro EXV_DEBUG could be defined as **//** (in-line comment) by default and std::cout when DEBUG is set.  

Sometime we write to std::cerr, and sometimes to std::cout.  I prefer std::cout, however let's be consistent.

**Another cosmetic suggestion**

I'm wondering out loud.  Maybe this isn't a good idea!

Can we use a vector/switch to make the code look simpler.  Instead of `if ( this and that ) { ... } else { foo man choo } ...` Something like:
```
    Exiv2::StringVector menu;
    menu.add("zTXt"); int c_zTxt = 0;
    menu.add("iTXt"); int c_iTXt = 1;
    switch  ( menu.find(chunkType) ) {
      case c_zTxt : 
      break;
    }

**Comment by piponazo** [comment]
@clanmills You comments make total sense and I would be happy to work on them. However, would you mind if we take care about that in other branch? I think that the changes would not be related with this branch which is fixing #597.

We can create another Github issue and place your comments there so that we do not forget about them.

**Comment by clanmills** [comment]
Sure, that's fine.  Can you fix the thread safety issue?  Let's open another issue for the cosmetic stuff.

**Pull request merged**